python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -0,0 +1,270 @@
+# Copyright (c) 2020-2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import math
+import pickle
+from typing import Any, Dict, Mapping, Sequence, Tuple, Type, TypeVar
+
+import numpy
+
+import rmm
+
+import cudf
+from cudf.core.abc import Serializable
+from cudf.utils.string import format_bytes
+
+T = TypeVar("T", bound="Buffer")
+
+
+class Buffer(Serializable):
+    """A Buffer represents device memory.
+
+    Usually the factory function `as_buffer` should be used to
+    create a Buffer instance.
+
+    Parameters
+    ----------
+    ptr : int
+        An integer representing a pointer to device memory.
+    size : int
+        Size of device memory in bytes.
+    owner : object
+        Python object to which the lifetime of the memory allocation is tied.
+    """
+
+    _ptr: int
+    _size: int
+    _owner: object
+
+    def __init__(self, ptr: int, size: int, owner: object):
+        if size < 0:
```

Comment:
Actually I'm proposing something even more radical: change `Buffer`'s constructor to raise an Exception. The factories can do `cls.__new__(cls)` and then initialize members (looks like only really necessary in `from_device_memory`). I think both `from_device_memory` and `from_host_memory` should be internal (prefixed with an underscore). The _only_ publicly available method for creating `Buffer`s should be `as_buffer` (the factories are available for internal use, aka inside `as_buffer`).

Reply:
Done